sdk/python/kfp/compiler/_data_passing_rewriter.py

```diff
@@ -381,3 +383,337 @@ def deconstruct_single_placeholder(s: str) -> List[str]:
     if not re.fullmatch('{{[-._a-zA-Z0-9]+}}', s):
         return None
     return s.lstrip('{').rstrip('}').split('.')
+
+
+def _replace_output_dir_placeholder(command_line: str,
+    output_directory: Optional[str] = None) -> str:
+    """Replaces the output directory placeholder."""
+    if _components.OUTPUT_DIR_PLACEHOLDER in command_line:
+        if not output_directory:
+            raise ValueError('output_directory of a pipeline must be specified '
+                             'when URI placeholder is used.')
+        return command_line.replace(
+            _components.OUTPUT_DIR_PLACEHOLDER, output_directory)
+    return command_line
+
+
+def _refactor_outputs_if_uri_placeholder(
+    container_template: Dict[str, Any],
+    output_to_filename: Dict[str, str]
+) -> None:
+    """Rewrites the output of the container in case of URI placeholder.
+
+    Also, collects the mapping from the output names to output file name.
+
+    Args:
+        container_template: The container template structure.
+        output_to_filename: The mapping from the artifact name to the actual file
+            name of the content. This will be used later when reconciling the
+            URIs on the consumer side.
+    """
+
+    # If there's no artifact outputs then no refactor is needed.
+    if not container_template.get('outputs') or not container_template[
+        'outputs'].get('artifacts'):
+        return
+
+    parameter_outputs = container_template['outputs'].get('parameters') or []
+    new_artifact_outputs = []
+    for artifact_output in container_template['outputs']['artifacts']:
+        # Check if this is an output associated with URI placeholder based
+        # on its path.
+        if _components.OUTPUT_DIR_PLACEHOLDER in artifact_output['path']:
+            # If so, we'll add a parameter output to output the pod name
+            parameter_outputs.append(
+                {
+                    'name': _components.PRODUCER_POD_NAME_PARAMETER.format(
+                        artifact_output['name']),
+                    'value': '{{pod.name}}'
+                })
+            output_to_filename[artifact_output['name']] = os.path.basename(
+                artifact_output['path'])
```

Comment:
So output URI is determined by initially generated `artifact_output['path']` path?
I remember it always generates path ending with `data`?

Is there a way users can specify file name for the artifact? e.g. if the artifact is an image, usually many image related tools expect the file name to end with a certain suffix representing image type.

Reply:
> So output URI is determined by initially generated artifact_output['path'] path?
I remember it always generates path ending with data?

Yes it's assumed to be `artifact_output['path']`. However for URI-based placeholder it's ending in the producer's corresponding output name, instead of `data`.

> Is there a way users can specify file name for the artifact?

There is no support at this moment, but users can still implement such contract in their user logic. For example, the producer component writes a file ending in a specific extension under the designated URI location, and the consumer read from that location.

Reply:
For file artifacts it's useful to have /data at the end. This way a single artifact can be mounted to a directory.